hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/utils/Pipelines.java

```diff
@@ -450,4 +473,159 @@ public static class DummySink implements SinkFunction<Object> {
+      tablePath = conf.get(FlinkOptions.PATH);
+      fileSystem = FileSystem.get(URI.create(tablePath));
+      checkpointLock = new Object();
+      String[] partitionFields = conf.get(FlinkOptions.PARTITION_PATH_FIELD).split(",");
+      partitionKeys = Arrays.asList(partitionFields);
+      partitionSuccessFileDelay = conf.get(FlinkOptions.PARTITION_WRITE_SUCCESS_FILE_DELAY);
+      String partitionValueExtractorClzName = conf.get(FlinkOptions.HIVE_SYNC_PARTITION_EXTRACTOR_CLASS_NAME);
+      String partitionTimestampExtractPattern = conf.get(FlinkOptions.PARTITION_TIME_EXTRACTOR_TIMESTAMP_PATTERN);
+      String partitionTimestampFormatPattern = conf.get(FlinkOptions.PARTITION_TIME_EXTRACTOR_TIMESTAMP_FORMATTER);
+      try {
+        Class<?> partitionValueExtractorClz = Class.forName(partitionValueExtractorClzName);
+        partitionValueExtractor = (PartitionValueExtractor) partitionValueExtractorClz.newInstance();
+      } catch (ClassNotFoundException e) {
+        LOG.error("class not found for: {}", partitionValueExtractorClzName, e);
+        throw e;
+      }
+      partitionTimeExtractor = new PartitionTimeExtractorAdapter(partitionTimestampExtractPattern, partitionTimestampFormatPattern);
+    }
+
+    // Extract the partition time value from partition path, and convert them to timestamp.
+    private Long convertTimestampByPartitionPath(String partitionPath) {
+      List<String> partitionValues = partitionValueExtractor.extractPartitionValuesInPath(partitionPath);
+      LocalDateTime localDateTime = partitionTimeExtractor.extract(partitionKeys, partitionValues);
+      return PartitionTimeExtractorAdapter.toMills(localDateTime);
+    }
+
+    @Override
+    public void invoke(Object value, SinkFunction.Context context) throws Exception {
+      String partition = (String) value;
+      activePartitions.add(partition);
+      long watermark = context.currentWatermark();
+      Iterator<String> it = activePartitions.iterator();
+      while (it.hasNext()) {
+        String partitionPath = it.next();
+        // Convert the partition path to timestamp if the table is partitioned by time field, like day, hour
+        Long partitionTimestamp = convertTimestampByPartitionPath(partitionPath);
+        // If the watermark is greater than the partition timestamp plus the delay time, it represents the
+        // minimum timestamp in the streaming data is beyond the partition max timestamp, so add the partition
+        // path to the finished partitions set and remove it from active partitions set.
+        if (partitionTimestamp + partitionSuccessFileDelay.toMillis() < watermark) {
+          synchronized (checkpointLock) {
+            finishedPartitions.add(partitionPath);
+            it.remove();
+          }
+        }
+      }
+    }
+
+    @Override
+    public void snapshotState(FunctionSnapshotContext functionSnapshotContext) throws Exception {
+      synchronized (checkpointLock) {
```

Comment:
This means that when multiple writer threads write, do they need to isolate the state?

Reply:
The streaming task and the checkpoint timer may concurrent modify the finished partition set , so I add a lock to provide thread safety. But according to the mailbox feature of flink version above 1.10,  the checkpoint lock is not need now, so I remove this lock. And thanks for your review.